### PR TITLE
feat(ec2.py): enhance regions input

### PR DESCRIPTION
**Pull request by hanks**
##### SUMMARY

Enhance regions input like below:

1. add `--regions` option
   currently there is no option for regions, so add this for handy
2. support space and comma for region input
   because in ec2.ini, input for `regions_exclude` uses `comma and space`, but `regions` can not, it is a little confused, Add this to make them consistent
3. fix logic to always exclude regions specified
   seems `regions_exclude` does not be used for all the conditions, so add logic to exclude every time
4. fix logic to call eucalyptus, there is no usage like `self.regions.append(boto.connect_euca(host=self.eucalyptus_host).region.name, **self.credentials)`, shoule be `self.regions.append(boto.connect_euca(host=self.eucalyptus_host, **self.credentials).region.name)`
5. add unittest for regions part in ec2.py
6. add docs for options usage
7. refactor a little for unit test easily

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

contrib/inventory/ec2.py

##### ANSIBLE VERSION

```paste below
ansible 2.7.0
  config file = None
  configured module search path = ['/Users/hanzhou/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/var/pyenv/versions/3.6.4/envs/3.6.4/lib/python3.6/site-packages/ansible
  executable location = /usr/local/var/pyenv/versions/3.6.4/bin/ansible
  python version = 3.6.4 (default, Mar 23 2018, 19:03:30) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)]
```

### CODE REVIEW

File: test/units/contrib/inventory/test_ec2_inventory.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+import os
+import sys
+import unittest
+from unittest.mock import MagicMock, patch
```
Comment (mattclay) [inline test/units/contrib/inventory/test_ec2_inventory.py]: New unit tests should be written to use pytest instead of unittest.

**Comment by mattclay** [comment]
CI failure in unit tests: https://app.shippable.com/github/ansible/ansible/runs/87358/4/tests

**Comment by hanks** [comment]
@mattclay thank you for the review

ok, I will use pytest, and for the unittest error, because I use the built in libs in python3.
Do I need to make the test run both py3 and py2, or just py3?  

Thanks for your time

**Comment by mattclay** [comment]
@hanks The unit tests need to run on Python versions 2.6, 2.7, 3.5, 3.6 and 3.7 currently.

**Comment by hanks** [comment]
@mattclay I changed to pytest follow docs from [https://docs.ansible.com/ansible/devel/dev_guide/testing_units.html](https://docs.ansible.com/ansible/devel/dev_guide/testing_units.html), Please check.

And one issue is that I met like below:

```
08:33 ------------------------------- Captured stderr --------------------------------
08:33 Failed to import docker-py. Try `pip install docker-py` - cannot import name 'APIClient' from 'docker' (/root/ansible/contrib/inventory/docker.py)
08:33 __ ERROR collecting test/units/modules/cloud/docker/test_docker_container.py ___
08:33 test/units/modules/cloud/docker/test_docker_container.py:3: in <module>
08:33     from ansible.modules.cloud.docker.docker_container import TaskParameters
08:33 lib/ansible/modules/cloud/docker/docker_container.py:710: in <module>
08:33     from ansible.module_utils.docker_common import (
08:33 lib/ansible/module_utils/docker_common.py:34: in <module>
08:33     from docker import __version__ as docker_version
08:33 contrib/inventory/docker.py:893: in <module>
08:33     main()
08:33 contrib/inventory/docker.py:888: in main
08:33     fail("Failed to import docker-py. Try `pip install docker-py` - %s" % (HAS_DOCKER_ERROR))
08:33 contrib/inventory/docker.py:422: in fail
08:33     sys.exit(1)
08:33 E   SystemExit: 1
```

seems there are not enough unit tests for `crontrib/inventory`, and when run my test code, it will import the top-level `docker-py` from `crontrib/inventory/docker.py`, So I also added into units requirements.txt

File: test/units/contrib/inventory/test_ec2_inventory.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+# contrib's dirstruct doesn't contain __init__.py files
+# import ec2 inventory manually
+root_dir = os.path.dirname(
+    os.path.dirname(
+        os.path.dirname(
+            os.path.dirname(os.path.dirname(os.path.abspath(__file__))))))
+inventory_dir = os.path.join(root_dir, 'contrib', 'inventory')
+inventory_dir_abspath = os.path.abspath(inventory_dir)
+sys.path.append(inventory_dir_abspath)
```
Comment (mattclay) [inline test/units/contrib/inventory/test_ec2_inventory.py]: It's not safe to modify sys.path outside of unit test functions, as these changes will affect other unrelated tests.

Comment (mattclay) [reply test/units/contrib/inventory/test_ec2_inventory.py]: You may be able to modify sys.path, perform an import, and then restore sys.path. That will probably prevent it from causing issues for other unit tests, but I haven't verified that.

Comment (hanks) [reply test/units/contrib/inventory/test_ec2_inventory.py]: yeah, now I use `teardown_module` in `pytest` to restore sys.path.

```python
def teardown_module(module):
    # cleanup so that path is not polluted with other scripts
    if inventory_dir_abspath in sys.path:
        sys.path.remove(inventory_dir_abspath)
```

Comment (hanks) [reply test/units/contrib/inventory/test_ec2_inventory.py]: now I use `teardown_module` to restore `sys.path`, and as you mentioned, it seems to be not safe, so fix like below:

```python
sys.path.append(inventory_dir_abspath)
from ec2 import Ec2Inventory
# cleanup so that path is not polluted with other scripts
sys.path.remove(inventory_dir_abspath)
```

File: test/runner/requirements/units.txt [hunk 3]
```diff
@@ -1,4 +1,6 @@
+boto
 boto3
+docker-py
```
Comment (mattclay) [inline test/runner/requirements/units.txt]: This import shouldn't be needed with the sys.path being cleaned up.

File: test/units/contrib/inventory/test_ec2_inventory.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+def mock_connect_euca_call(mocker):
+    regionMock = mocker.MagicMock()
+    regionMock.name = 'us-gov-west-1'
+    mock1 = mocker.MagicMock(region=regionMock)
+
+    return mocker.patch(
+        'ec2.boto.connect_euca', mocker.MagicMock(return_value=mock1))
+
+
+class TestEC2Inventory(object):
```
Comment (mattclay) [inline test/units/contrib/inventory/test_ec2_inventory.py]: When using pytest, create top-level functions without using a class.

File: test/units/contrib/inventory/test_ec2_inventory.py (hunk 2 above)
Comment (mattclay) [reply test/units/contrib/inventory/test_ec2_inventory.py]: I found a better way to handle the import, using a pytest fixture.

I have a PR (https://github.com/ansible/ansible/pull/46732) to fix the vmware_inventory unit tests that provides a good example:

https://github.com/ansible/ansible/pull/46732/files#diff-7a31f0c775c6683a8212d1f175db50e9R28

The `BaseException` handling was only needed because the inventory script uses `sys.exit` for import failures.

Comment (hanks) [reply test/units/contrib/inventory/test_ec2_inventory.py]: cool! fixed as your way

File: test/units/contrib/inventory/test_ec2_inventory.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    inventory_dir = os.path.abspath(
+        os.path.join(
+            os.path.dirname(__file__), '..', '..', '..', '..', 'contrib',
+            'inventory'))
+    sys.path.append(inventory_dir)
+
+    try:
+        import ec2
+    except ImportError as ex:
+        pytest.fail(ex)
```
Comment (mattclay) [inline test/units/contrib/inventory/test_ec2_inventory.py]: Since only `ImportError` is being handled here, use this instead:

```python
try:
    ec2 = pytest.importorskip("ec2")
finally:
    # rest of code here
```

That way the tests will be properly skipped if the import isn't available.

Comment (hanks) [reply test/units/contrib/inventory/test_ec2_inventory.py]: fixed it, so the policy is like to skip test if `ImportError` happens, and do not fail it?

Comment (mattclay) [reply test/units/contrib/inventory/test_ec2_inventory.py]: Correct.

Comment (hanks) [reply test/units/contrib/inventory/test_ec2_inventory.py]: Hi, is there anything improvement I can do for this PR?

**Comment by hanks** [comment]
@mattclay Hi, is there anything improvement I can do for this PR?

File: .gitignore [hunk 6]
```diff
@@ -66,6 +66,7 @@ coverage.xml
 /test/units/cover-html
 /test/integration/targets/*/backup/
 /test/cache/*
+metadata-*.json
```
Comment (mattclay) [inline .gitignore:69]: This change should be made in a separate PR since it's not directly related to the rest of the changes here.

**Comment by mattclay** [comment]
@hanks Thanks for including tests with your changes. Have you tried finding someone in the Ansible community that uses this inventory script to test and review your changes? I recommend asking in the #ansible-aws IRC channel on Freenode.

**Comment by s-hertel** [comment]
Hi @hanks, nice that you included tests for this. :smile: We aren't working on adding features to the contrib scripts since we have added inventory plugins. The aws_ec2 inventory plugin uses boto3, has more flexible group and hostvar configuration, has common caching infrastructure, and uses configuration files rather than passing command line arguments. Have you looked into the aws_ec2 inventory plugin to see if it meets your needs? https://docs.ansible.com/ansible/latest/plugins/inventory.html#using-inventory-plugins

If it does not, submitting changes to modify https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/inventory/aws_ec2.py is a better way to make changes to EC2 inventory at this point. However, since bugfixes may still be added to contrib/inventory/ec2.py, having unit tests would be really nice. Do you think there's a point to separating tests from your changes?
